File: zedmini_prob_grid.cpp
// Minimal standalone example: probability grid mapping with ZED Mini + OpenCV.
// Dependencies: ZED SDK, OpenCV (core/imgproc/highgui). Build example:
//   g++ -std=c++17 zedmini_prob_grid.cpp -o zedmini_prob_grid \ 
//       -I/usr/local/zed/include -L/usr/local/zed/lib -lsl_zed \
//       `pkg-config --cflags --libs opencv4`
// Run near-field only; adjust grid size/resolution for your field dimensions.

#include <sl/Camera.hpp>
#include <opencv2/opencv.hpp>
#include <vector>
#include <limits>
#include <cmath>
#include <iostream>

struct Grid {
  int w, h;
  float res; // meters per cell
  std::vector<float> log_odds;
  float l_hit = 0.7f, l_miss = -0.4f, l_decay = 0.98f;

  Grid(int w_, int h_, float res_) : w(w_), h(h_), res(res_), log_odds(w_ * h_, 0.0f) {}
  int idx(int x, int y) const { return y * w + x; }
  void decay() { for (auto &v : log_odds) v *= l_decay; }
  void updateCell(int x, int y, bool hit) {
    if (x < 0 || y < 0 || x >= w || y >= h) return;
    log_odds[idx(x, y)] += hit ? l_hit : l_miss;
  }
  bool bestCell(int &bx, int &by, float &score) const {
    score = -std::numeric_limits<float>::infinity();
    bool found = false;
    for (int y = 0; y < h; ++y) {
      for (int x = 0; x < w; ++x) {
        float v = log_odds[idx(x, y)];
        if (v > score) {
          score = v;
          bx = x;
          by = y;
          found = true;
        }
      }
    }
    return found;
  }
};

bool detectBall(const cv::Mat &bgr, cv::Point &px_center) {
  cv::Mat hsv, mask;
  cv::cvtColor(bgr, hsv, cv::COLOR_BGR2HSV);
  // Example threshold for orange ball; tune for lighting/ball color.
  cv::inRange(hsv, cv::Scalar(5, 120, 120), cv::Scalar(25, 255, 255), mask);
  std::vector<std::vector<cv::Point>> cnts;
  cv::findContours(mask, cnts, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
  static constexpr double MIN_BALL_AREA = 80.0;
  double bestArea = 0;
  cv::Point best(0, 0);
  for (auto &c : cnts) {
    double a = cv::contourArea(c);
    if (a > MIN_BALL_AREA && a > bestArea) {
      bestArea = a;
      cv::Moments m = cv::moments(c);
      if (m.m00 != 0) {
        best = {int(m.m10 / m.m00), int(m.m01 / m.m00)};
      }
    }
  }
  if (bestArea == 0) return false;
  px_center = best;
  return true;
}

int main() {
  static constexpr float MIN_VALID_DEPTH = 0.2f;
  static constexpr float MAX_VALID_DEPTH = 8.0f;
  static constexpr float CAMERA_HEIGHT = 0.45f;      // meters above ground
  static constexpr float CAMERA_PITCH_RAD = 0.0f;    // downward-positive pitch
  static constexpr float MIN_RAY_DOWNWARD_COMPONENT = 1e-3f;
  static constexpr int DRAW_RADIUS = 6;
  static constexpr int DRAW_THICKNESS = 2;
  const cv::Scalar DRAW_COLOR(0, 255, 0);

  sl::Camera zed;
  sl::InitParameters p;
  p.camera_resolution = sl::RESOLUTION::HD720;
  p.depth_mode = sl::DEPTH_MODE::PERFORMANCE; // faster, lower depth accuracy; switch to sl::DEPTH_MODE::ULTRA if precision is needed
  p.coordinate_units = sl::UNIT::METER;
  p.coordinate_system = sl::COORDINATE_SYSTEM::RIGHT_HANDED_Z_UP_X_FWD;
  if (zed.open(p) != sl::ERROR_CODE::SUCCESS) {
    std::cerr << "Failed to open ZED\n";
    return 1;
  }

  auto calib = zed.getCameraInformation().camera_configuration.calibration_parameters.left_cam;
  float fx = calib.fx, fy = calib.fy, cx = calib.cx, cy = calib.cy;

  // 10m x 10m area centered laterally; forward is +Z.
  Grid grid(200, 200, 0.05f);
  sl::Mat zedImage, zedDepth;

  std::cout << "Press ESC to quit.\n";
  while (true) {
    if (zed.grab() != sl::ERROR_CODE::SUCCESS) continue;
    zed.retrieveImage(zedImage, sl::VIEW::LEFT);
    zed.retrieveMeasure(zedDepth, sl::MEASURE::DEPTH);

    cv::Mat rgba(zedImage.getHeight(), zedImage.getWidth(), CV_8UC4, zedImage.getPtr<sl::uchar1>(sl::MEM::CPU));
    cv::Mat bgr;
    cv::cvtColor(rgba, bgr, cv::COLOR_BGRA2BGR);

    grid.decay();
    cv::Point px;
    bool seen = detectBall(bgr, px);
    if (seen) {
      float depth = 0.0f;
      sl::ERROR_CODE err = zedDepth.getValue(px.x, px.y, &depth);
      if (err == sl::ERROR_CODE::SUCCESS && std::isfinite(depth) && depth > MIN_VALID_DEPTH && depth < MAX_VALID_DEPTH) {
        float X = (px.x - cx) * depth / fx; // right
        float Y = (px.y - cy) * depth / fy; // down
        float Z = depth;                    // forward

        // Apply pitch about camera X (right) axis to account for mounting angle.
        float cosP = std::cos(CAMERA_PITCH_RAD);
        float sinP = std::sin(CAMERA_PITCH_RAD);
        float Yr = cosP * Y - sinP * Z;
        float Zr = sinP * Y + cosP * Z;
        float Xr = X;

        // Intersect ray with ground plane at camera height; skip if ray is parallel or pointing upward.
        if (Yr > MIN_RAY_DOWNWARD_COMPONENT) {
          float scale = CAMERA_HEIGHT / Yr;
          float ground_x = Xr * scale; // lateral
          float ground_z = Zr * scale; // forward
          int gx = int(ground_x / grid.res) + grid.w / 2;
          int gy = int(ground_z / grid.res);
          grid.updateCell(gx, gy, true);
        }
        cv::circle(bgr, px, DRAW_RADIUS, DRAW_COLOR, DRAW_THICKNESS);
      }
    }

    int bx = 0, by = 0;
    float score = 0;
    if (grid.bestCell(bx, by, score)) {
      float target_x = (bx - grid.w / 2) * grid.res;
      float target_z = by * grid.res;
      std::cout << "Target cell (" << target_x << "m lateral, " << target_z
                << "m forward), score=" << score << "\n";
    }

    cv::imshow("ZED Ball Detection", bgr);
    if (cv::waitKey(1) == 27) break; // ESC
  }

  zed.close();
  return 0;
}

File: zedmini_prob_grid.py
"""
Minimal Python example: ZED Mini probability grid mapping with OpenCV.
Dependencies:
  - ZED SDK Python API (pyzed-sl)
  - OpenCV (`pip install opencv-python`)

Run:
  python zedmini_prob_grid.py
"""

import math
import numpy as np
import cv2
import pyzed.sl as sl


class Grid:
  def __init__(self, w: int, h: int, res: float):
    self.w, self.h, self.res = w, h, res
    self.log_odds = np.zeros((h, w), dtype=np.float32)
    self.l_hit, self.l_miss, self.l_decay = 0.7, -0.4, 0.98

  def decay(self):
    self.log_odds *= self.l_decay

  def update(self, x: int, y: int, hit: bool):
    if 0 <= x < self.w and 0 <= y < self.h:
      self.log_odds[y, x] += self.l_hit if hit else self.l_miss

  def best(self):
    idx = np.unravel_index(np.argmax(self.log_odds), self.log_odds.shape)
    score = self.log_odds[idx]
    return (idx[1], idx[0], float(score))  # (x, y, score)


def _find_contours(mask):
  # OpenCV 4 returns (contours, hierarchy); OpenCV 3 returns (image, contours, hierarchy)
  out = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
  return out[1] if len(out) == 3 else out[0]


def detect_ball(bgr: np.ndarray):
  hsv = cv2.cvtColor(bgr, cv2.COLOR_BGR2HSV)
  mask = cv2.inRange(hsv, (5, 120, 120), (25, 255, 255))  # orange example
  cnts = _find_contours(mask)
  MIN_BALL_AREA = 80.0
  best_area = 0.0
  best = None
  for c in cnts:
    area = cv2.contourArea(c)
    if area > MIN_BALL_AREA and area > best_area:
      m = cv2.moments(c)
      if m["m00"] != 0:
        best_area = area
        best = (int(m["m10"] / m["m00"]), int(m["m01"] / m["m00"]))
  return best


def is_valid_depth(depth_value, error_code, min_depth, max_depth):
  return error_code == sl.ERROR_CODE.SUCCESS and math.isfinite(depth_value) and min_depth < depth_value < max_depth


def main():
  MIN_VALID_DEPTH = 0.2
  MAX_VALID_DEPTH = 8.0
  CAMERA_HEIGHT = 0.45  # meters
  CAMERA_PITCH_RAD = 0.0  # downward-positive: +X rotation in RIGHT_HANDED_Z_UP_X_FWD tilts camera down
  MIN_RAY_DOWN = 1e-3
  DRAW_RADIUS, DRAW_THICKNESS = 6, 2
  DRAW_COLOR = (0, 255, 0)

  grid = Grid(200, 200, 0.05)  # 10m x 10m

  zed = sl.Camera()
  init_params = sl.InitParameters(
      camera_resolution=sl.RESOLUTION.HD720,
      depth_mode=sl.DEPTH_MODE.PERFORMANCE,  # faster; use ULTRA for higher accuracy
      coordinate_units=sl.UNIT.METER,
      coordinate_system=sl.COORDINATE_SYSTEM.RIGHT_HANDED_Z_UP_X_FWD
  )
  if zed.open(init_params) != sl.ERROR_CODE.SUCCESS:
    print("Failed to open ZED")
    return

  cam_info = zed.get_camera_information()
  calib = cam_info.camera_configuration.calibration_parameters.left_cam
  fx, fy, cx, cy = calib.fx, calib.fy, calib.cx, calib.cy

  image = sl.Mat()
  depth = sl.Mat()

  print("Press ESC to quit.")
  while True:
    if zed.grab() != sl.ERROR_CODE.SUCCESS:
      continue

    zed.retrieve_image(image, sl.VIEW.LEFT)
    zed.retrieve_measure(depth, sl.MEASURE.DEPTH)

    bgr = image.get_data()[:, :, :3].copy()
    grid.decay()

    px = detect_ball(bgr)
    if px:
      depth_value, error_code = depth.get_value(px[0], px[1])
      if is_valid_depth(depth_value, error_code, MIN_VALID_DEPTH, MAX_VALID_DEPTH):
        x_cam = (px[0] - cx) * depth_value / fx
        y_cam = (px[1] - cy) * depth_value / fy
        z_cam = depth_value

        cos_p, sin_p = math.cos(CAMERA_PITCH_RAD), math.sin(CAMERA_PITCH_RAD)
        y_r = cos_p * y_cam - sin_p * z_cam
        z_r = sin_p * y_cam + cos_p * z_cam
        x_r = x_cam

        if y_r > MIN_RAY_DOWN:
          scale = CAMERA_HEIGHT / y_r
          gx = int((x_r * scale) / grid.res) + grid.w // 2
          gy = int((z_r * scale) / grid.res)
          grid.update(gx, gy, True)
        cv2.circle(bgr, px, DRAW_RADIUS, DRAW_COLOR, DRAW_THICKNESS)

    bx, by, score = grid.best()
    target_x = (bx - grid.w // 2) * grid.res
    target_z = by * grid.res
    msg = f"Target: {target_x:.2f}m, {target_z:.2f}m, s={score:.2f}"
    print(msg)
    cv2.putText(bgr, msg, (20, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2)

    cv2.imshow("ZED Ball Detection", bgr)
    if cv2.waitKey(1) == 27:
      break

  zed.close()


if __name__ == "__main__":
  main()
